[LTC] Refactor torch_lazy_tensors::Device

alanwaketan opened the pull request:
Summary:
This pull request will keep track of the progress of refactoring torch_lazy_tensors::Device.
1. The first commit removed torch_lazy_tensors::DeviceType from all places and added a HardwareType type as an extension point for backends that need this type of information.
2. Clean up class Device interface.
3. Remove most use cases of string devices.
4. Introduce BackendDeviceType to class Device.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

To be noted:
It's easier to review it commit by commit.

Commit:
[LTC] Remove torch_lazy_tensors::DeviceType

Summary:
This commit removed torch_lazy_tensors::DeviceType from all places and
added a HardwareType type as an extension point for backends that need
this type of information.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

Commit:
[LTC] Clean up class Device interface

Summary:
This commit cleans up the interfaces of class Device a bit.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

Commit:
[LTC] Remove most use cases of string devices

Summary:
This commit replaces most of the string devices with Device devices, and
leaves a TODO on BackendImplInterface.GetLocalDevices().

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

Commit:
[LTC] Introduce BackendDeviceType to class Device

Summary:
This aiming of this new BackendDeviceType is to provide an extensible
way for backends to pretty print device type while keeping the many-to-one
design philophy of class Device intact and avoid changing too many code.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

alanwaketan commented on lazy_tensor_core/lazy_tensor_core/csrc/device.h:
```diff
@@ -1,50 +1,47 @@
 
-enum class DeviceType { CPU, GPU, TPU };
+// Backend can extend it and define their own supported hardware types.
+struct BackendDeviceType {
```
@JackCaoG Here is the new design. This aiming of this new BackendDeviceType is to provide an extensible way for backends to pretty print device type while keeping the many-to-one design philosophy of class Device intact and avoid changing too many code.

Why we need to pretty print Device? It's because we have (and want to offer) APIs that allow users to query the current hardware.

alanwaketan reviewed (commented):
The pull request is ready for review.

wconstab reviewed (approved):
Thanks @alanwaketan looks great.  Let's wait for @JackCaoG to review.

For naming, Do you think it makes sense to use `BackendDevice`, `BackendData` (my other PR), `BackendInterface`?  I wanted a way to make it very clear these generic sounding objects are from the backend rather than the similar ones in pytorch..  Other option might be putting them in a `backend:: namespace`.

alanwaketan commented:
> Thanks @alanwaketan looks great. Let's wait for @JackCaoG to review.
> 
> For naming, Do you think it makes sense to use `BackendDevice`, `BackendData` (my other PR), `BackendInterface`? I wanted a way to make it very clear these generic sounding objects are from the backend rather than the similar ones in pytorch.. Other option might be putting them in a `backend:: namespace`.

I think Backend* types make sense, and should probably stay in torch::lazy. The reason why I didn't make the name change is that I recognize that I need to do the renaming again after merging master. So, I would just leave the renaming work then.

Commit:
[LTC] Remove string device from LazyGraphExecutor::ScheduleSyncTensorsGraph

Summary:
This commit removes string device from LazyGraphExecutor::ScheduleSyncTensorsGraph.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

JackCaoG reviewed (commented):
Mostly LGTM

alanwaketan commented on lazy_tensor_core/lazy_tensor_core/csrc/device.h:
```diff
@@ -1,50 +1,47 @@
-  }
-
-  std::string ToString() const;
+  std::string toString() const;
```
FB style...

alanwaketan commented on lazy_tensor_core/lazy_tensor_core/csrc/ir_dump_util.h:
```diff
@@ -2,10 +2,12 @@
 
 namespace torch_lazy_tensors {
+
+class Device;
```
Using forward declarations instead of include in headers can help reduce build time.

alanwaketan commented on lazy_tensor_core/lazy_tensor_core/csrc/lazy_graph_executor.cpp:
```diff
@@ -436,17 +436,11 @@ void LazyGraphExecutor::MarkStep(const Device& device) {
 }
 
-void LazyGraphExecutor::WaitDeviceOps(c10::ArrayRef<std::string> devices) {
+void LazyGraphExecutor::WaitDeviceOps() {
```
We currently haven't finalized the backend API to do string -> device mapping yet. Let me add a TODO and change the signatures to take `c10::ArrayRef<Device>`.

wconstab replied:
Yea, you can just leave a TODO for now and I can add this back in with my PR that changes the backend interfaces.  Or whatever is easiest for you.

JackCaoG reviewed (approved):
Thanks!

Commit:
[LTC] Revert part of 67f9f544

Summary:
Using `using namespace ...` in headers will break the build in unexpected
ways.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

Commit:
[LTC] Add back devices to LazyGraphExecutor::WaitDeviceOps

Summary:
This commit adds back devices to LazyGraphExecutor::WaitDeviceOps as such
user API is useful.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

alanwaketan commented:
@wconstab @JackCaoG Thanks for approving the pull request. I will merge it once the CI passes.
